crates/pallets/starknet/src/blockifier_state_adapter.rs

```diff
@@ -92,17 +94,53 @@ impl<T: Config> State for BlockifierStateAdapter<T> {
         self.storage_update.insert((contract_address, key), value);
         let contract_address: ContractAddressWrapper = contract_address.0.0.into();
         let key: StorageKeyWrapper = key.0.0.into();
+        let value = Felt252Wrapper::from(value);
 
         let contract_storage_key: ContractStorageKeyWrapper = (contract_address, key);
 
-        crate::StorageView::<T>::insert(contract_storage_key, Felt252Wrapper::from(value));
+        crate::StorageView::<T>::insert(contract_storage_key, value);
+
+        // Update state tries if enabled in the runtime configuration
+        if T::EnableStateRoot::get() {
+            // Update contracts trie
+            let mut tree = crate::State::<T>::get().storage_commitment;
+            let nonce = Pallet::<T>::nonce(contract_address);
+            let class_hash = Pallet::<T>::contract_class_hash_by_address(contract_address).unwrap_or_default();
+
+            // Update storage trie
+            let mut state_tree = crate::State::<T>::get().state_commitment;
+            state_tree.set(key, value);
+            let state_root = state_tree.commit();
```

Comment:
We should only commit in the finalize block. Here everytime we write in the storage or do something we commit which is extremely long. same for every function in the blockifier adapter

Reply:
But do we really need to update the leaf at each transaction?
Can't we just push values in the tree at each tx and only commit at the very end?

Reply:
That's what we do, apart from the state root which is necessary to compute the leaf hash for contracts tree

Reply:
hmm no we update every leaf of the contract tree during the block execution and then when we close a block we recompute the whole starknet tree to update the state root. We should compute the tree only once per block